File: apps/mobile/lib/data/models/enums.dart
/// Domain enums mirroring `packages/shared/src/enums.ts`.
///
/// Each parses from the wire string defensively: an unrecognised value falls
/// back to a sensible default rather than throwing.
library;

enum UserRole {
  student('STUDENT'),
  teacher('TEACHER'),
  instituteAdmin('INSTITUTE_ADMIN'),
  superAdmin('SUPER_ADMIN');

  const UserRole(this.wire);
  final String wire;

  static UserRole parse(String? value) => UserRole.values.firstWhere(
        (UserRole r) => r.wire == value,
        orElse: () => UserRole.student,
      );

  bool get isStaff => this == instituteAdmin || this == superAdmin;
}

enum VerificationStatus {
  unverified('UNVERIFIED'),
  pending('PENDING'),
  verified('VERIFIED'),
  rejected('REJECTED');

  const VerificationStatus(this.wire);
  final String wire;

  static VerificationStatus parse(String? value) =>
      VerificationStatus.values.firstWhere(
        (VerificationStatus s) => s.wire == value,
        orElse: () => VerificationStatus.unverified,
      );

  bool get isVerified => this == verified;
}

enum CourseType {
  online('ONLINE', 'Online'),
  inPerson('IN_PERSON', 'In person'),
  hybrid('HYBRID', 'Hybrid');

  const CourseType(this.wire, this.label);
  final String wire;
  final String label;

  static CourseType parse(String? value) => CourseType.values.firstWhere(
        (CourseType t) => t.wire == value,
        orElse: () => CourseType.inPerson,
      );
}

enum CourseLevel {
  beginner('BEGINNER', 'Beginner'),
  intermediate('INTERMEDIATE', 'Intermediate'),
  advanced('ADVANCED', 'Advanced'),
  allLevels('ALL_LEVELS', 'All levels');

  const CourseLevel(this.wire, this.label);
  final String wire;
  final String label;

  static CourseLevel parse(String? value) => CourseLevel.values.firstWhere(
        (CourseLevel l) => l.wire == value,
        orElse: () => CourseLevel.allLevels,
      );
}

enum LeadStatus {
  newLead('NEW', 'New'),
  contacted('CONTACTED', 'Contacted'),
  interviewed('INTERVIEWED', 'Interviewed'),
  enrolled('ENROLLED', 'Enrolled'),
  cancelled('CANCELLED', 'Cancelled');

  const LeadStatus(this.wire, this.label);
  final String wire;
  final String label;

  static LeadStatus parse(String? value) => LeadStatus.values.firstWhere(
        (LeadStatus s) => s.wire == value,
        orElse: () => LeadStatus.newLead,
      );
}

enum FormFieldType {
  text('TEXT'),
  textarea('TEXTAREA'),
  number('NUMBER'),
  email('EMAIL'),
  phone('PHONE'),
  date('DATE'),
  select('SELECT'),
  multiSelect('MULTI_SELECT'),
  checkbox('CHECKBOX'),
  file('FILE'),
  nationalId('NATIONAL_ID');

  const FormFieldType(this.wire);
  final String wire;

  static FormFieldType parse(String? value) => FormFieldType.values.firstWhere(
        (FormFieldType t) => t.wire == value,
        orElse: () => FormFieldType.text,
      );
}

enum QuestionType {
  multipleChoice('MULTIPLE_CHOICE'),
  multiSelect('MULTI_SELECT'),
  trueFalse('TRUE_FALSE'),
  shortAnswer('SHORT_ANSWER'),
  essay('ESSAY'),
  fileUpload('FILE_UPLOAD');

  const QuestionType(this.wire);
  final String wire;

  static QuestionType parse(String? value) => QuestionType.values.firstWhere(
        (QuestionType t) => t.wire == value,
        orElse: () => QuestionType.multipleChoice,
      );

  /// True when the student picks from a fixed option list.
  bool get isChoice =>
      this == multipleChoice || this == multiSelect || this == trueFalse;

  bool get isText => this == shortAnswer || this == essay;
}

enum AttemptStatus {
  inProgress('IN_PROGRESS'),
  submitted('SUBMITTED'),
  autoSubmitted('AUTO_SUBMITTED'),
  graded('GRADED'),
  voided('VOIDED');

  const AttemptStatus(this.wire);
  final String wire;

  static AttemptStatus parse(String? value) => AttemptStatus.values.firstWhere(
        (AttemptStatus s) => s.wire == value,
        orElse: () => AttemptStatus.inProgress,
      );

  bool get isOpen => this == inProgress;
}

enum EnrollmentStatus {
  active('ACTIVE', 'Active'),
  completed('COMPLETED', 'Completed'),
  dropped('DROPPED', 'Dropped'),
  suspended('SUSPENDED', 'Suspended');

  const EnrollmentStatus(this.wire, this.label);
  final String wire;
  final String label;

  static EnrollmentStatus parse(String? value) =>
      EnrollmentStatus.values.firstWhere(
        (EnrollmentStatus s) => s.wire == value,
        orElse: () => EnrollmentStatus.active,
      );
}

enum MediaKind {
  image('IMAGE'),
  video('VIDEO'),
  audio('AUDIO'),
  document('DOCUMENT'),
  panorama360('PANORAMA_360');

  const MediaKind(this.wire);
  final String wire;

  static MediaKind parse(String? value) => MediaKind.values.firstWhere(
        (MediaKind k) => k.wire == value,
        orElse: () => MediaKind.document,
      );
}

enum LiveClassProvider {
  adobeConnect('ADOBE_CONNECT', 'Adobe Connect'),
  bigBlueButton('BIG_BLUE_BUTTON', 'BigBlueButton');

  const LiveClassProvider(this.wire, this.label);
  final String wire;
  final String label;

  static LiveClassProvider parse(String? value) =>
      LiveClassProvider.values.firstWhere(
        (LiveClassProvider p) => p.wire == value,
        orElse: () => LiveClassProvider.bigBlueButton,
      );
}

enum DiscoverySort {
  distance('distance', 'Nearest'),
  rating('rating', 'Top rated'),
  price('price', 'Lowest price'),
  popularity('popularity', 'Most popular');

  const DiscoverySort(this.wire, this.label);
  final String wire;
  final String label;
}

File: apps/api/src/packages/shared/src/enums.ts
/**
 * Domain enumerations shared between the NestJS API and the Next.js clients.
 * Keep these in sync with `apps/api/prisma/schema.prisma`.
 */

export const UserRole = {
  STUDENT: 'STUDENT',
  TEACHER: 'TEACHER',
  INSTITUTE_ADMIN: 'INSTITUTE_ADMIN',
  SUPER_ADMIN: 'SUPER_ADMIN',
} as const;
export type UserRole = (typeof UserRole)[keyof typeof UserRole];

export const VerificationStatus = {
  UNVERIFIED: 'UNVERIFIED',
  PENDING: 'PENDING',
  VERIFIED: 'VERIFIED',
  REJECTED: 'REJECTED',
} as const;
export type VerificationStatus =
  (typeof VerificationStatus)[keyof typeof VerificationStatus];

export const CourseType = {
  ONLINE: 'ONLINE',
  IN_PERSON: 'IN_PERSON',
  HYBRID: 'HYBRID',
} as const;
export type CourseType = (typeof CourseType)[keyof typeof CourseType];

export const CourseLevel = {
  BEGINNER: 'BEGINNER',
  INTERMEDIATE: 'INTERMEDIATE',
  ADVANCED: 'ADVANCED',
  ALL_LEVELS: 'ALL_LEVELS',
} as const;
export type CourseLevel = (typeof CourseLevel)[keyof typeof CourseLevel];

/** Kanban columns for the institute CRM lead board. */
export const LeadStatus = {
  NEW: 'NEW',
  CONTACTED: 'CONTACTED',
  INTERVIEWED: 'INTERVIEWED',
  ENROLLED: 'ENROLLED',
  CANCELLED: 'CANCELLED',
} as const;
export type LeadStatus = (typeof LeadStatus)[keyof typeof LeadStatus];

export const LEAD_STATUS_ORDER: LeadStatus[] = [
  LeadStatus.NEW,
  LeadStatus.CONTACTED,
  LeadStatus.INTERVIEWED,
  LeadStatus.ENROLLED,
  LeadStatus.CANCELLED,
];

/**
 * Allowed lead transitions. A lead may always be cancelled, may move one step
 * forward, or be pulled one step back (except out of a terminal state).
 */
export const LEAD_TRANSITIONS: Record<LeadStatus, LeadStatus[]> = {
  NEW: ['CONTACTED', 'CANCELLED'],
  CONTACTED: ['INTERVIEWED', 'ENROLLED', 'NEW', 'CANCELLED'],
  INTERVIEWED: ['ENROLLED', 'CONTACTED', 'CANCELLED'],
  ENROLLED: ['CANCELLED'],
  CANCELLED: ['NEW'],
};

export const FormFieldType = {
  TEXT: 'TEXT',
  TEXTAREA: 'TEXTAREA',
  NUMBER: 'NUMBER',
  EMAIL: 'EMAIL',
  PHONE: 'PHONE',
  DATE: 'DATE',
  SELECT: 'SELECT',
  MULTI_SELECT: 'MULTI_SELECT',
  CHECKBOX: 'CHECKBOX',
  FILE: 'FILE',
  NATIONAL_ID: 'NATIONAL_ID',
} as const;
export type FormFieldType = (typeof FormFieldType)[keyof typeof FormFieldType];

export const QuestionType = {
  MULTIPLE_CHOICE: 'MULTIPLE_CHOICE',
  MULTI_SELECT: 'MULTI_SELECT',
  TRUE_FALSE: 'TRUE_FALSE',
  SHORT_ANSWER: 'SHORT_ANSWER',
  ESSAY: 'ESSAY',
  FILE_UPLOAD: 'FILE_UPLOAD',
} as const;
export type QuestionType = (typeof QuestionType)[keyof typeof QuestionType];

/** Question types the engine can grade without a human in the loop. */
export const AUTO_GRADED_TYPES: QuestionType[] = [
  QuestionType.MULTIPLE_CHOICE,
  QuestionType.MULTI_SELECT,
  QuestionType.TRUE_FALSE,
  QuestionType.SHORT_ANSWER,
];

export const AttemptStatus = {
  IN_PROGRESS: 'IN_PROGRESS',
  SUBMITTED: 'SUBMITTED',
  AUTO_SUBMITTED: 'AUTO_SUBMITTED',
  GRADED: 'GRADED',
  VOIDED: 'VOIDED',
} as const;
export type AttemptStatus = (typeof AttemptStatus)[keyof typeof AttemptStatus];

export const EnrollmentStatus = {
  ACTIVE: 'ACTIVE',
  COMPLETED: 'COMPLETED',
  DROPPED: 'DROPPED',
  SUSPENDED: 'SUSPENDED',
} as const;
export type EnrollmentStatus =
  (typeof EnrollmentStatus)[keyof typeof EnrollmentStatus];

export const TransactionType = {
  ENROLLMENT_REVENUE: 'ENROLLMENT_REVENUE',
  PLATFORM_COMMISSION: 'PLATFORM_COMMISSION',
  PAYOUT: 'PAYOUT',
  REFUND: 'REFUND',
  ADJUSTMENT: 'ADJUSTMENT',
} as const;
export type TransactionType =
  (typeof TransactionType)[keyof typeof TransactionType];

export const PayoutStatus = {
  NONE: 'NONE',
  REQUESTED: 'REQUESTED',
  APPROVED: 'APPROVED',
  PAID: 'PAID',
  REJECTED: 'REJECTED',
} as const;
export type PayoutStatus = (typeof PayoutStatus)[keyof typeof PayoutStatus];

export const MediaKind = {
  IMAGE: 'IMAGE',
  VIDEO: 'VIDEO',
  AUDIO: 'AUDIO',
  DOCUMENT: 'DOCUMENT',
  PANORAMA_360: 'PANORAMA_360',
} as const;
export type MediaKind = (typeof MediaKind)[keyof typeof MediaKind];

export const MediaStatus = {
  PENDING_UPLOAD: 'PENDING_UPLOAD',
  UPLOADED: 'UPLOADED',
  PROCESSING: 'PROCESSING',
  READY: 'READY',
  FAILED: 'FAILED',
} as const;
export type MediaStatus = (typeof MediaStatus)[keyof typeof MediaStatus];

export const LiveClassProvider = {
  ADOBE_CONNECT: 'ADOBE_CONNECT',
  BIG_BLUE_BUTTON: 'BIG_BLUE_BUTTON',
} as const;
export type LiveClassProvider =
  (typeof LiveClassProvider)[keyof typeof LiveClassProvider];

export const BookingStatus = {
  AVAILABLE: 'AVAILABLE',
  BOOKED: 'BOOKED',
  CANCELLED: 'CANCELLED',
  COMPLETED: 'COMPLETED',
} as const;
export type BookingStatus = (typeof BookingStatus)[keyof typeof BookingStatus];

/** Categories used for map pin icons and discovery filters. */
export const CATEGORY_SLUGS = [
  'languages',
  'programming',
  'music',
  'arts',
  'academic-tutoring',
  'test-prep',
  'business',
  'sports',
] as const;
export type CategorySlug = (typeof CATEGORY_SLUGS)[number];
